pkg/bundles/kontext/bundle.go

```diff
@@ -44,16 +46,51 @@ func bundle(directory string) (v1.Layer, error) {
 	tw := tar.NewWriter(buf)
 	defer tw.Close()
 
+	excludedDirs := sets.NewString()
+	includesDirs := sets.NewString()
+
 	err := filepath.Walk(directory,
 		func(path string, fi os.FileInfo, err error) error {
 			if err != nil {
 				return err
 			}
 
-			// Skip anything in the .git directory
-			// TODO(mattmoor): expand this to .gitignore / .dockerignore?
-			if fi.IsDir() && filepath.Base(path) == ".git" {
-				return filepath.SkipDir
+			//Always include the root directory
+			includesDirs.Insert(directory)
+			if directory != path {
+				//if parent is included then dont check them children, include them implicitly
+				var isParentIncluded bool
+
+				if fi.IsDir() {
+					isParentIncluded = includesDirs.Has(path)
+				} else {
+					//As parent is always included make sure this applies only to sub directories
+					isParentIncluded = includesDirs.Has(filepath.Dir(path)) && filepath.Dir(path) != directory
+				}
+
+				if !isParentIncluded {
+					//Check if the directory has  .dockerignore
+					ignorer, err := ignore.NewOrDefault(directory)
```

Comment:
So `directory` is the root of the walk, and doesn't change, but we're setting this up for every file we walk. 🤔 

I'd expect (above the `filepath.Walk`):
```
ignorer, err := ignore.NewOrDefault(directory)
if err != nil {
   return err
}
```

Then in the `filepath.Walk` we should simply have:
```
if ignorer.Ignore(relativePath) {
    return filepath.SkipDir
}
```

The goal should be to hide everything else behind the ignorer abstraction.

Reply:
See my comment on the other PR, but we should refine the interface to support skipping directories or individual files.  we can use something like:
```go
switch ignorer.Ignore(relativePath) {
case ignore.File: return nil
case ignore.Dir: return filepath.SkipDir
default: // fall through
}
```

Reply:
> So `directory` is the root of the walk, and doesn't change, but we're setting this up for every file we walk. 🤔
> 
> I'd expect (above the `filepath.Walk`):
> 
> ```
> ignorer, err := ignore.NewOrDefault(directory)
> if err != nil {
>    return err
> }
> ```
> 
That makes total sense
> Then in the `filepath.Walk` we should simply have:
> 
> ```
> if ignorer.Ignore(relativePath) {
>     return filepath.SkipDir
> }
> ```
> 
Not sure I got this comment right? SkipDir skips the direcotry so this has to be done only if `fileInfo.IsDir()` ??
> The goal should be to hide everything else behind the ignorer abstraction.

Reply:
Yeah, I peeked at the code, and unfortunately Go isn't relaxed about `filepath.SkipDir` being returned for files 😞 
```
			err = walk(filename, fileInfo, walkFn)
			if err != nil {
				if !fileInfo.IsDir() || err != SkipDir {
					return err
				}
			}
```

This might mean we should pass the `os.FileMode` into the `Ignore` callback as well, but I still think the enum return is probably our best bet.

Reply:
> See my comment on the other PR, but we should refine the interface to support skipping directories or individual files. we can use something like:
> 
> ```go
> switch ignorer.Ignore(relativePath) {
> case ignore.File: return nil
> case ignore.Dir: return filepath.SkipDir
I still feel this might break if the we have inversions under the directory 
e.g say we want to  skip `target` directroy but I dont want to skip `target/foo-runner.jar`, then skipDir will never traverse target and we might miss the other includes.  The typical example `.dockeringore` for this case will be like:
```
target
!target/*-runner.jar
```
> default: // fall through
> }
> ```

Reply:
Yeah, this is why in slack I said File/Dir were sort of misnomers and changed to Current/Transitive (and talked about passing file mode):
```
switch ignorer.Ignore(relativePath, info.Mode()) {
case ignore.Current: return nil
case ignore.Transitive: return filepath.SkipDir
}
```

In this scheme my expectation would be that the ignorer would know that `target` encloses an inversion so `ignore.Transitive` is inappropriate and therefore return `ignore.Current`.